### DISCUSSION TITLE
how to add timer to loop_node

### DISCUSSION THREAD

**Discussion by melisa-akan** [jspsych/jsPsych#1067] [Q&A] [answered]
Hello! 

As you see below, I have a "loop_node" (var loop_trials) and I want it to run until a pre-determined amount of time has elapsed. Is this at all possible? I want the "timer" to start when loop_trials is first pushed to the timeline. When XX ms have elapsed, I want to see the next thing in the timeline.  

I would really appreciate any guidance.

Thanks a lot for all your help!  

`
var loop_trials = {
  timeline: [rdk_fixation, rdk_trial, response_trial, feedback_trial],
  loop_function: function(){
	if (time_elapsed == XX){ //if XX ms have elapse since the start of loop_trials, end it
	return false;
	}else{
	return true;
	}
  
  }
 
};
`

**Comment by becky-gilbert** [comment, marked as answer]
Yes. I would use a global variable to hold the start time, then inside your loop function you can calculate the elapsed time based on the start time.
Note that in the if statement, I changed the comparison to _greater than_ or equal to XX, since it's very unlikely that the elapsed time will be equal to your threshold time when the loop function runs.
```javascript
var start_time = performance.now()
var loop_trials = {
  timeline: [rdk_fixation, rdk_trial, response_trial, feedback_trial],
  loop_function: function(){
    var time_elapsed = performance.now() - start_time;
    if (time_elapsed >= XX){ //if XX ms have elapse since the start of loop_trials, end it
      return false;
    }else{
      return true;
    }
  }
};
```
Another thing to note is that the code above starts the timer when the page first loads. If that's not what you want then you could instead create the `start_time` variable without giving it a value (so that it's still a global variable), and then set it to `performance.now()` somewhere else, such as when the first trial starts, using an [event-related callback function](https://www.jspsych.org/overview/callbacks/).

**Comment by melisa-akan** [comment]
I don't know how to thank you @becky-gilbert! Your help means so much! If I am able to complete my dissertation studies, it will be thanks to you and @jodeleeuw! Really grateful!

**Reply by becky-gilbert** [reply]
No problem!
